sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -736,6 +737,24 @@ abstract class TypeCoercionBase {
     }
   }
 
+  /**
+   * Determines the value type of a [[Melt]].
+   */
+  object MeltCoercion extends Rule[LogicalPlan] {
+    override def apply(plan: LogicalPlan): LogicalPlan =
+      plan resolveOperators {
+        case m: Melt if m.values.nonEmpty && m.values.forall(_.resolved) && m.valueType.isEmpty =>
```

Comment:
A more common way is to add cast to value expressions to make them the same type. e.g.
```
case m: Melt if m.values.nonEmpty && !haveSameType(m.values.map(_.dataType)) => {
  val commonType = ...
  if (commonType.isDefined) {
    m.copy(values = values.map(Cast(_, commonType.get)))
  } else {
    fail...
  }
}
```

Reply:
Another point is, usually we don't fail in analyzer rules. If we can't resolve it, we leave the plan unchanged and fail in `CheckAnalysis`

Reply:
This cast is done in `Analyzer.scala` in identical way. You are saying it is preferred to move this into `TypeCoercion`?

Reply:
I think `m.values.forall(_.resolved)` is still needed, otherwise `if m.values.nonEmpty && !haveSameType(m.values.map(_.dataType))` could call this rule when `values` are not resolved and `m.values.map(_.dataType)` would throw an exception.

Reply:
Cast moved into `MeltCoercion` in 477a42f7c19dc9fd29bd3c2f9e1bc5a5532731c0.